sycl/doc/extensions/experimental/sycl_ext_oneapi_device_architecture.asciidoc

```diff
@@ -959,6 +962,9 @@ struct architecture;
 _Return type:_ `sycl::ext::oneapi::experimental::architecture`
 
 _Returns:_ The architecture of the device.
+
+If current device architecture is not supported by the extension, a
+`sycl::exception` with error code `sycl::errc::runtime` will be thrown.
```

Comment:
In this case, I think we could add a new enumeration `architecture::unknown` and return that.

Reply:
If we add `architecture::unknown`, we can have the following situation:

- the architecture is really not supported by the extension
- and also - when we have a bug and the current device architecture should be supported, but for some reason not

So for the second bullet the query returns the same `architecture::unknown` and we can't detect that bug appears.

Reply:
Can we distinguish between the two cases today?  In the current code, do we throw an exception in both cases?

Reply:
Yes, we throw the exception anyway, because both bullets use the same API: `device::<ext::oneapi::experimental::info::device::architecture>`.

In short, I'm ok to change to `architecture::unknown` but

* it will not reduce the number of false positives we receive (I'm mentioning the bug with Critical exposure we received yesterday)
* customers one way or another should check the result of the query: with the exception they have to catch this exception, and with `architecture::unknown` returned, they should compare the result with `architecture::unknown`, which yes, quicker then catching the exception
* our tests will not benefit from having `architecture::unknown`: their behavior will be completely the same - we will fail if `architecture::unknown` returns

Reply:
I understand what you are saying, but I still think it is worth adding `architecture::unknown` for this case.  As we saw recently, people don't expect these query APIs to throw an exception and they are surprised when an exception causes their application to terminate.  I think many common use cases will "just work" if we return an error code.  For example:

```
syclex::architecture arch = d.get_info<syclex::info::device::architecture>();
switch (arch) {
  case syclex::architecture::x86_64:
    /* do something */
  case syclex::architecture::intel_gpu_pvc:
    /* do something else */
  default:
    /* generic code that works on all devices */
}
```

I think this would be a common code pattern, and it will continue to work even if `get_info` returns `architecture::unknown`.

Inside device code, calls like this should also work as expected:

```
syclex::if_architecture_is<syclex::architecture::x86_64>([&]{
  // Do something
}).else_if_architecture_ge<syclex::architecture::intel_gpu_pvc>([&]{
  // Do something else
}).otherwise([&]{
  // Generic code
});
```

Note that I would expect `device::ext_oneapi_architecture_is(syclex::architecture::unknown)` to return `true` if the device does not have a known architecture.  I doubt much code will do this, but this seems like the logical behavior

I suspect we will not be able to make `syclex::if_architecture_is<syclex::architecture::unknown>` execute the "true" lamdba with our current AOT implementation.  However, I think we can make this work once we implement the [full design](https://github.com/intel/llvm/pull/9127) that supports JIT.